Add changeToLayoutContainer action and is-layout-container attribute to amp-list for resizing

cathyxz opened the pull request:
Experimentally addresses #10230. 

This addresses the resizable-children issue of giving developers a way to trigger `changeToLayoutContainer_` on an event or on bind, currently via the `changeToLayoutContainer` action and `is-layout-container` bindable attribute. As is, it will directly call `changeToLayoutContainer`, which undos the original layout with caveats (e.g. preserves fixed width for layout fixed), and apply layout container to the amp-list. This will allow the amp-list to automatically resize based on the size of its children after the layout container change. Repeated calls of the `changeToLayoutContainer` action will have no effect. 

If `changeToLayoutContainer` is called while the overflow element is visible and present, it will change to layout container and cause the overflow element to disappear. 


- [x] Testing on more devices and browsers
- [x] Action and attribute naming discussion
- [x] ~Unit tests~
- [x] Integration tests

aghassemi commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -134,6 +134,10 @@ export class AmpList extends AMP.BaseElement {
       }
     }, ActionTrust.HIGH);
 
+    this.registerAction('containerize',
```
let's guard the registration around the experiment flag

cathyxz replied:
Done

aghassemi commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -134,6 +134,10 @@ export class AmpList extends AMP.BaseElement {
 
+    this.registerAction('containerize',
+        () => this.changeToLayoutContainer_(),
+        ActionTrust.LOW);
```
must be `HIGH` trust.

cathyxz replied:
Done

aghassemi commented on extensions/amp-bind/0.1/bind-impl.js:
```diff
@@ -72,7 +72,7 @@ let BoundElementDef;
 const BIND_ONLY_ATTRIBUTES = map({
   'AMP-CAROUSEL': ['slide'],
-  'AMP-LIST': ['state'],
+  'AMP-LIST': ['state', 'is-container'],
```
keep a todo or issue somewhere to add this to validator protoascii file before removing experiment flag.

cathyxz commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -249,6 +244,9 @@ export class AmpList extends AMP.BaseElement {
       promise = this.scheduleRender_(data, /*append*/ false);
     }
+    if (isContainer) {
+      this.changeToLayoutContainer_();
```
Good point! It has no effect if set to `false` because changing to layout container is a one-way road (at least right now).

cathyxz replied:
Great reminder to add documentation.

cathyxz commented on extensions/amp-list/amp-list.md:
````diff
@@ -183,6 +183,23 @@ The `amp-list` element exposes a `refresh` action that other elements can refere
 </amp-list>
 ```
 
````
/cc @CrystalFaith for documentation changes.

aghassemi commented on extensions/amp-list/amp-list.md:
```diff
@@ -249,6 +266,9 @@ Displays a loading indicator and placeholder again when the list's source is ref
 
 By default, this will only trigger on refreshes that cause a network fetch. To reset on all refreshes, use `reset-on-refresh="always"`.
 
+#### is-layout-container (experimental, optional)
```
`[is-layout-container]`

Please also explain in the doc why this is only bindable and not available as a normal attribute (e.g. user action is needed for container since reflow, etc...)

cathyxz replied:
Done

aghassemi replied:
put it in `[]` since it is a bind-only attribute.

alabiaga commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -642,30 +630,38 @@ export class AmpList extends AMP.BaseElement {
-    this.element.setAttribute('layout', 'container');
+
+    return this.mutateElement(() => {
+      this.undoPreviousLayout_(previousLayout);
```
undoPreviousLayout_ switch statements needs a default case even if it's a fall through, based on js style guide.

default:
  // fall through.

https://google.github.io/styleguide/jsguide.html#features-switch-statements

cathyxz replied:
Good to know. Thank you!

cathyxz commented:
Some of the lint fixes were breaking tests so I reverted them / amended the gulp lint fix commit. Fixed a naming bug in the integration tests for this feature. Should now be good to go.
